include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+    void *ptr = free_va_.get(size, alignment);
+    if (ptr)
+      return ptr;
+    va_allocate(size);
+    ptr = free_va_.get(size, alignment);
+    assert(ptr);
+    return ptr;
+  }
+
+  /**
+   * @brief Allocate virtual address space of at at least `min_size` bytes.
+   *
+   * The function hints the driver to use a distinct address space for each device in the
+   * attempt to have a contiguous address spaces for each device. Currently, the spacing
+   * between device VA spaces is 1 TiB and initial VA size for each device is 4 GiB.
+   */
+  void va_allocate(size_t min_size) {
+    size_t va_size = std::max(next_pow2(min_size), initial_va_size_);
+
+    CUdeviceptr hint = {};
+    va_region *last_va = nullptr;
+    if (va_regions_.empty()) {
+      // Calculate the hint address for the allocations for this device.
+      // There's some not very significant base and different devices get
+      // address spaces separated by 2^40 - this should be quite enough.
+      hint = (device_ordinal_ + 1) * 0x10000000000u;
+    } else {
+      // Try to allocate after the last VA for this device
+      last_va = &va_regions_.back();
+      hint = last_va->address_range.end();
+    }
+    va_ranges_.push_back(cuvm::CUMemAddressRange::Reserve(va_size, 0, hint));
+    auto &va = va_ranges_.back();
+
+    va_add_region(va);
+    stat_va_add(va_size);
+  }
+
+  /**
+   * @brief Add a memory region that spans the given VA range and merge it with adjacent
+   *        regions, if found.
+   */
+  void va_add_region(cuvm::CUAddressRange va) {
+    // Try to merge regions
+    // 1. Find preceding region
+    va_region *region = nullptr;
+    for (auto &r : va_regions_)
+      if (r.address_range.end() == va.ptr()) {
+        region = &r;
+        break;
+      }
+    if (region) {
+      // Found! Resize it. The new VA range is unmapped, so we can simply resize,
+      // there's nothing more to do.
+      region->resize(region->num_blocks() + va.size() / block_size_);
+    } else {
+      // Not found - create a new region
+      va_regions_.emplace_back(va, block_size_);
+      region = &va_regions_.back();
+    }
+
+    // 2. Find succeeding region
+    for (size_t i = 0; i < va_regions_.size(); i++) {
+      auto &r = va_regions_[i];
+      if (r.address_range.ptr() == region->address_range.end()) {
+        // Found - now we append the old range (which follows the new one) to the new one...
+        region->append(std::move(r));
+        // ...and remove the old one
+        va_regions_.erase_at(i);
+        break;
+      }
+    }
+
+    // 3. Place the virtual address range in the VA free tree
+    free_va_.put(reinterpret_cast<void*>(va.ptr()), va.size());
+  }
+
+  void do_deallocate(void *ptr, size_t size, size_t alignment) override {
+    deallocate_impl(ptr, size, alignment, true);
+  }
+
+  void deallocate_impl(void *ptr, size_t size, size_t alignment, bool sync) {
+    adjust_params(size, alignment);
+    if (size == 0)
```

Comment:
Can you move to the begining of the function?

Reply:
Done.